packages/roosterjs-editor-core/lib/undo/Undo.ts

```diff
@@ -0,0 +1,226 @@
+                this.onKeyPress(event as PluginDomEvent);
+                break;
+            case PluginEventType.CompositionEnd:
+                this.clearRedoForInput();
+                break;
+            case PluginEventType.ContentChanged:
+                if ((<ContentChangedEvent>event).source != 'Undo' && !this.isRestoring) {
+                    this.clearRedoForInput();
+                }
+                break;
+        }
+    }
+
+    /**
+     * Clear all existing undo snapshots
+     */
+    public clear() {
+        this.undoSnapshots = null;
+        this.hasNewContent = false;
+    }
+
+    /**
+     * Restore an undo snapshot to editor
+     */
+    public undo(): void {
+        if (this.hasNewContent) {
+            this.addUndoSnapshot();
+        }
+
+        this.restoreSnapshot(-1 /*previousSnapshot*/);
+    }
+
+    /**
+     * Restore a redo snapshot to editor
+     */
+    public redo(): void {
+        this.restoreSnapshot(1 /*nextSnapshot*/);
+    }
+
+    /**
+     * Whether there is a snapshot for undo
+     */
+    public canUndo(): boolean {
+        return this.hasNewContent || this.getSnapshotsManager().canMove(-1 /*previousSnapshot*/);
+    }
+
+    /**
+     * Whether there is a snapshot for redo
+     */
+    public canRedo(): boolean {
+        return this.getSnapshotsManager().canMove(1 /*nextSnapshot*/);
+    }
+
+    /**
+     * Add an undo snapshot
+     */
+    public addUndoSnapshot(): void {
+        let snapshot = buildSnapshot(this.editor);
+        this.getSnapshotsManager().addSnapshot(snapshot);
+        this.hasNewContent = false;
+    }
+
+    private restoreSnapshot(delta: number) {
+        let snapshot = this.getSnapshotsManager().move(delta);
+
+        if (snapshot) {
+            try {
+                this.isRestoring = true;
+                restoreSnapshot(this.editor, snapshot);
+            } finally {
+                this.isRestoring = false;
+            }
+
+            this.fireContentChangedEvent();
+        }
+    }
+
+    private onKeyDown(pluginEvent: PluginDomEvent): void {
+        // Handle backspace/delete when there is a selection to take a snapshot
+        // since we want the state prior to deletion restorable
+        let evt = pluginEvent.rawEvent as KeyboardEvent;
+        let shouldTakeUndo = false;
+        if (evt.which == KEY_BACKSPACE || evt.which == KEY_DELETE) {
+            let selectionRange = this.editor.getSelectionRange();
+            if (
+                selectionRange &&
+                (!selectionRange.collapsed ||
+                    // If the selection contains image, we need to add undo snapshots
+                    containsImage(selectionRange.startContainer))
+            ) {
+                shouldTakeUndo = true;
+            }
+        } else if (this.hasNewContent && evt.which >= KEY_PAGEUP && evt.which <= KEY_DOWN) {
+            // PageUp, PageDown, Home, End, Left, Right, Up, Down
+            shouldTakeUndo = true;
```

Comment:
It may add a lot of snapshots if user keep press "Left"

Reply:
Wouldn't the this.hasNewContent solve that? As long as that's getting set to false after a snapshot is created, then this doesn't matter.

It's also only checking for PageUp and PageDown right now. Is that intentional?

Reply:
I guess the other keys are >= KEY_PageUp and <= KEY_DOWN, :D

Reply:
Lego is right. In that case hasNewContent will be false.

And Bo is right too, these 8 keys are in the range of 33-40.